OpenRA.Mods.Common/Traits/World/ControlGroups.cs

```diff
@@ -0,0 +1,144 @@
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class ControlGroupsInfo : TraitInfo, IControlGroupsInfo
+	{
+		public readonly string[] Groups = { "1", "2", "3", "4", "5", "6", "7", "8", "9", "0" };
+
+		public override object Create(ActorInitializer init) { return new ControlGroups(init.World, this); }
+
+		string[] IControlGroupsInfo.Groups => Groups;
+	}
+
+	[TraitLocation(SystemActors.World | SystemActors.EditorWorld)]
+	public class ControlGroups : IControlGroups, ITick, IGameSaveTraitData
+	{
+		readonly World world;
+		public string[] Groups { get; private set; }
+
+		readonly List<Actor>[] controlGroups;
+
+		public ControlGroups(World world, ControlGroupsInfo info)
+		{
+			this.world = world;
+			Groups = info.Groups;
+			controlGroups = Enumerable.Range(0, Groups.Length).Select(_ => new List<Actor>()).ToArray();
+		}
+
+		public void SelectControlGroup(int group)
+		{
+			world.Selection.Combine(world, GetActorsInControlGroup(group), false, false);
+		}
+
+		public void CreateControlGroup(int group)
+		{
+			if (!world.Selection.Actors.Any())
+				return;
+
+			controlGroups[group].Clear();
+
+			RemoveActorsFromAllControlGroups(world.Selection.Actors);
+
+			controlGroups[group].AddRange(world.Selection.Actors.Where(a => a.Owner == world.LocalPlayer));
+		}
+
+		public void AddSelectionToControlGroup(int group)
+		{
+			if (!world.Selection.Actors.Any())
+				return;
+
+			RemoveActorsFromAllControlGroups(world.Selection.Actors);
+
+			controlGroups[group].AddRange(world.Selection.Actors.Where(a => a.Owner == world.LocalPlayer));
+		}
+
+		public void CombineSelectionWithControlGroup(int group)
+		{
+			world.Selection.Combine(world, GetActorsInControlGroup(group), true, false);
+		}
+
+		public void AddToControlGroup(Actor a, int group)
+		{
+			if (!controlGroups[group].Contains(a))
+				controlGroups[group].Add(a);
+		}
+
+		public void RemoveFromControlGroup(Actor a)
+		{
+			var group = GetControlGroupForActor(a);
+			if (group.HasValue)
+				controlGroups[group.Value].Remove(a);
+		}
+
+		public int? GetControlGroupForActor(Actor a)
+		{
+			for (var i = 0; i < controlGroups.Length; i++)
+				if (controlGroups[i].Contains(a))
+					return i;
+
+			return null;
+		}
+
+		void RemoveActorsFromAllControlGroups(IEnumerable<Actor> actors)
+		{
+			for (var i = 0; i < Groups.Length; i++)
+				controlGroups[i].RemoveAll(a => actors.Contains(a));
+		}
+
+		public IEnumerable<Actor> GetActorsInControlGroup(int group)
+		{
+			return controlGroups[group].Where(a => a.IsInWorld);
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			foreach (var cg in controlGroups)
+			{
+				// note: NOT `!a.IsInWorld`, since that would remove things that are in transports.
+				cg.RemoveAll(a => a.Disposed || a.Owner != world.LocalPlayer);
+			}
+		}
+
+		List<MiniYamlNode> IGameSaveTraitData.IssueTraitData(Actor self)
+		{
+			var groups = controlGroups
```

Comment:
This generates bogus data that causes a crash when loading a save.

Can fix with something like
```csharp
var groups = new List<MiniYamlNode>();
for (var i = 0; i < controlGroups.Length; i++)
{
	var cg = controlGroups[i];
	if (cg.Any())
	{
		var actorIds = cg.Select(a => a.ActorID).ToArray();
		groups.Add(new MiniYamlNode(i.ToString(), FieldSaver.FormatValue(actorIds)));
	}
}
```

Reply:
Oof, I completely glanced over this one. Fixed. Thanks.